### PR TITLE
feat: archive on delete support

### LINKED ISSUE #522: Implement archiveOnDelete feature which was provided by nfs-subdir-external-provisioner
**Is your feature request related to a problem?/Why is this needed**

I use both `csi-driver-nfs` and [kubernetes-sigs/nfs-subdir-external-provisioner](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner). I miss the feature from the subdir-external-provisioner that deleted volumes can be "archived" on the NFS share. Archive means renaming the root directory of the PV from `pvc-1234` to `archived-pvc-1234`.

**Describe the solution you'd like in detail**

It is implemented here:
https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner/blob/3740e016212145125285ee14667e7ab14e61b406/cmd/nfs-subdir-external-provisioner/provisioner.go#L172-L188

The archiving functionality is enabled when `parameters.archiveOnDelete` is set to `true` on the StorageClass:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
  labels:
    # ..
    chart: nfs-subdir-external-provisioner-4.0.17
  name: nfs-data
allowVolumeExpansion: true
parameters:
  archiveOnDelete: "true" # <------------------------------- see here
provisioner: cluster.local/nfs-data-provisioner
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

**Describe alternatives you've considered**

still use the [kubernetes-sigs/nfs-subdir-external-provisioner](https://github.com/kubernetes-sigs/nfs-subdir-external-provisioner) when archiving is needed.

**Additional context**

none

### PR DESCRIPTION
**Pull request by navilg**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>

/kind feature

**What this PR does / why we need it**:

This allows parameter `onDelete: archive` which will archive the the volume in nfs server when PV is deleted. Other two values, `delete` and `retain` are retained and serves their purpose.

This is helpful when we want to retain the nfs volume upon deletion of PV, but we also want to distinguish/separate the volume name in nfs server for deleted PVs from the in-use nfs volumes. This will prefix `archived-` in the folder name of PV in nfs server when PV is deleted.

**Which issue(s) this PR fixes**:

Fixes #522

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
To archive the volumes on deletion of PV, create storage class with parameter `ondelete: archive`. For e.g.

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: nfs-csi
mountOptions:
- nfsvers=4.1
parameters:
  onDelete: archive    # <----------------------------------------------------- Archive on delete policy
  server: mynfsserver.io
  share: /var/nfs/k8s_pvs
provisioner: nfs.csi.k8s.io
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer


To make the archive on deletion behaviour default, install with below value set in helm.

--set controller.defaultOnDeletePolicy=archive
```

### CODE REVIEW

File: pkg/nfs/controllerserver.go [hunk 1]
```diff
@@ -240,15 +240,27 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 			}
 		}()
 
-		// delete subdirectory under base-dir
 		internalVolumePath := getInternalVolumePath(cs.Driver.workingMountDir, nfsVol)
 
-		klog.V(2).Infof("Removing subdirectory at %v", internalVolumePath)
-		if err = os.RemoveAll(internalVolumePath); err != nil {
-			return nil, status.Errorf(codes.Internal, "failed to delete subdirectory: %v", err.Error())
+		if !strings.EqualFold(nfsVol.onDelete, archive) {
```
Comment (andyzhangx) [inline pkg/nfs/controllerserver.go:245]: the logic  like this is better:
```
if strings.EqualFold(nfsVol.onDelete, archive) {
 ...
} else {
 ...
}
```

**Review by andyzhangx — COMMENTED** [review]
/ok-to-test

**Comment by andyzhangx** [comment]
@navilg could you fix the ut failure? thanks.

**Comment by navilg** [comment]
@andyzhangx Thanks for fixing ut. I could not make time last weekend.

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**